src/meta-srv/src/handler/failure_handler.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod runner;
+
+use api::v1::meta::HeartbeatRequest;
+use async_trait::async_trait;
+
+use crate::error::Result;
+use crate::handler::failure_handler::runner::{FailureDetectControl, FailureDetectRunner};
+use crate::handler::{HeartbeatAccumulator, HeartbeatHandler};
+use crate::metasrv::{Context, ElectionRef};
+
+#[derive(Eq, Hash, PartialEq, Clone)]
+pub(crate) struct RegionIdent {
+    catalog: String,
+    schema: String,
+    table: String,
+    region_id: u64,
+}
+
+// TODO(LFC): TBC
+pub(crate) struct DatanodeHeartbeat {
+    #[allow(dead_code)]
+    cluster_id: u64,
+    #[allow(dead_code)]
+    node_id: u64,
+    region_idents: Vec<RegionIdent>,
+    heartbeat_time: i64,
+}
+
+pub struct RegionFailureHandler {
+    failure_detect_runner: FailureDetectRunner,
+}
+
+impl RegionFailureHandler {
+    pub fn new(election: Option<ElectionRef>) -> Self {
+        Self {
+            failure_detect_runner: FailureDetectRunner::new(election),
+        }
+    }
+
+    pub async fn start(&mut self) {
+        self.failure_detect_runner.start().await;
+    }
+}
+
+#[async_trait]
+impl HeartbeatHandler for RegionFailureHandler {
+    async fn handle(
+        &self,
+        _: &HeartbeatRequest,
+        ctx: &mut Context,
+        acc: &mut HeartbeatAccumulator,
+    ) -> Result<()> {
+        if ctx.is_infancy {
+            self.failure_detect_runner
+                .send_control(FailureDetectControl::Purge)
+                .await;
```

Comment:
`Purge` seems to require synchronous invocation?

Reply:
`Purge` command is sent through a mpsc channel, handled by a single task.